core/src/main/scala/ai/h2o/sparkling/backend/api/scalainterpreter/ScalaInterpreterServlet.scala

```diff
@@ -0,0 +1,188 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ai.h2o.sparkling.backend.api.scalainterpreter
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import ai.h2o.sparkling.backend.api.{DELETERequestBase, GETRequestBase, POSTRequestBase, ServletRegister}
+import ai.h2o.sparkling.backend.utils.{RestApiUtils, RestCommunication}
+import ai.h2o.sparkling.repl.H2OInterpreter
+import ai.h2o.sparkling.utils.SparkSessionUtils
+import javax.servlet.Servlet
+import javax.servlet.http.HttpServletRequest
+import org.apache.spark.h2o.H2OConf
+import water.api.schemas3.JobV3
+import water.exceptions.H2ONotFoundArgumentException
+
+import scala.collection.concurrent.TrieMap
+
+/**
+  * This servlet class handles requests for /3/scalaint endpoint
+  */
+private[api] class ScalaInterpreterServlet(conf: H2OConf)
+  extends GETRequestBase
+  with POSTRequestBase
+  with DELETERequestBase
+  with RestCommunication {
+
+  private val intrPoolSize = conf.scalaIntDefaultNum
+  private val freeInterpreters = new java.util.concurrent.ConcurrentLinkedQueue[H2OInterpreter]
+  private var mapIntr = new TrieMap[Int, H2OInterpreter]
+  private val lastIdUsed = new AtomicInteger(0)
+  private val jobResults = new TrieMap[String, ScalaCodeResult]
+  private val jobCount = new AtomicInteger(0)
+  initializeInterpreterPool()
+
+  def getSessions(): ScalaSessions = ScalaSessions(mapIntr.keys.toArray)
+
+  def interpret(sessionId: Int, code: String): ScalaCode = {
+    this.synchronized {
+      jobCount.incrementAndGet()
+      while (conf.maxParallelScalaCellJobs != -1 && jobCount
+               .intValue() > conf.maxParallelScalaCellJobs) {
+        Thread.sleep(1000)
```

Comment:
Why do we need `Thread.sleep` here?

Reply:
There are two scala cell execution modes -> synchronous and asynchronous. In synchronous we start the computation and get result right away. In asynchronous, the scala cell starts computing in a form on H2O Job and once H2O has result, Flow UI is able to interpret it. 

There was a customer requesting this feature some time ago. The sleep here is to ensure that we don't start more than `conf.maxParallelScalaCellJobs` parallel jobs in asynchronous mode

Reply:
Note: the sleep should no longer be needed, I have rewrote this piece of code using using executor service